src/lib/shim/src/syscall.rs

```diff
@@ -0,0 +1,229 @@
+                    return syscall_complete.retval;
+                }
+
+                if let Some(ctx) = ctx.as_mut() {
+                    // Set the syscall return value now, before potentially
+                    // invoking signal handlers. This appears to be the behavior
+                    // in the kernel; i.e. a handler for a signal that
+                    // is interrupted a blocking syscall should see the syscall
+                    // result (-EINTR) in the context passed to that handler.
+                    ctx.uc_mcontext.rax = syscall_complete.retval.into();
+                }
+
+                // SAFETY: `ctx` should be valid if present.
+                let all_sigactions_had_sa_restart =
+                    unsafe { crate::signals::process_signals(ctx.as_deref_mut()) };
+
+                if i64::from(syscall_complete.retval) == Errno::EINTR.to_negated_i64()
+                    && all_sigactions_had_sa_restart
+                    && syscall_complete.restartable
+                {
+                    // Restart syscall interrupted syscall
+                    crate::tls_ipc::with(|ipc| {
+                        ipc.to_shadow()
+                            .send(ShimEventToShadow::Syscall(*syscall_event))
+                    });
+                    continue;
+                } else {
+                    // Return syscall result
+                    return syscall_complete.retval;
+                }
+            }
+            ShimEventToShim::SyscallDoNative => {
+                let rv = unsafe { native_syscall(&syscall_event.syscall_args) };
+
+                if let FfiOption::Some(strace_fd) = crate::global_process_shmem::get().strace_fd {
+                    let emulated_time = global_host_shmem::get()
+                        .sim_time
+                        .load(atomic::Ordering::Relaxed)
+                        - EmulatedTime::SIMULATION_START;
+                    let tid = tls_thread_shmem::get().tid;
+                    let parts = TimeParts::from_nanos(emulated_time.as_nanos());
+                    let mut buffer = FormatBuffer::<200>::new();
+                    writeln!(
+                        &mut buffer,
+                        "{} [tid {}] ^^^ = {:?}",
+                        parts.fmt_hr_min_sec_nano(),
+                        tid,
+                        rv
+                    )
+                    .unwrap();
+                    // SAFETY: file descriptor should be valid and open.
+                    let strace_fd = unsafe { BorrowedFd::borrow_raw(strace_fd) };
+                    let mut strace_file_writer = BorrowedFdWriter::new(strace_fd);
+                    strace_file_writer.write_str(buffer.as_str()).unwrap();
+                }
+
+                return rv;
+            }
+            ShimEventToShim::Syscall(syscall) => {
+                let res = unsafe { native_syscall(&syscall.syscall_args) };
```

Comment:
Maybe add a comment about the difference here vs SyscallDoNative, since they are both calling `native_syscall`?

Reply:
done